#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""gRPC client for calling the bazi-core-service."""

from __future__ import annotations

import json
import logging
import os
import sys
from typing import Any, Dict, Optional

import grpc

# ÂØºÂÖ•ÁîüÊàêÁöÑ gRPC ‰ª£Á†Å
project_root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.insert(0, os.path.join(project_root, "proto", "generated"))

import bazi_core_pb2
import bazi_core_pb2_grpc

logger = logging.getLogger(__name__)


class BaziCoreClient:
    """gRPC client for the bazi-core-service."""

    def __init__(self, base_url: Optional[str] = None, timeout: float = 30.0) -> None:
        # base_url Ê†ºÂºè: host:port Êàñ [host]:port
        base_url = base_url or os.getenv("BAZI_CORE_SERVICE_URL", "")
        if not base_url:
            raise RuntimeError("BAZI_CORE_SERVICE_URL is not configured")
        
        # Ëß£ÊûêÂú∞ÂùÄÔºàÁßªÈô§ http:// ÂâçÁºÄÔºâ
        if base_url.startswith("http://"):
            base_url = base_url[7:]
        elif base_url.startswith("https://"):
            base_url = base_url[8:]
        
        # Â¶ÇÊûúÊ≤°ÊúâÁ´ØÂè£ÔºåÊ∑ªÂä†ÈªòËÆ§Á´ØÂè£
        if ":" not in base_url:
            base_url = f"{base_url}:9001"
        
        self.address = base_url
        self.timeout = timeout

    def calculate_bazi(self, solar_date: str, solar_time: str, gender: str) -> Dict[str, Any]:
        """ËÆ°ÁÆóÂÖ´Â≠óÊéíÁõò"""
        request = bazi_core_pb2.BaziCoreRequest(
            solar_date=solar_date,
            solar_time=solar_time,
            gender=gender,
        )

        import datetime
        request_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        logger.info(f"[{request_time}] üîµ Ë∞ÉÁî® bazi-core-service (gRPC): {self.address}, solar_date={solar_date}, solar_time={solar_time}, gender={gender}")
        logger.debug("Calling bazi-core-service (gRPC): %s request=%s", self.address, request)

        # ËÆæÁΩÆËøûÊé•ÈÄâÈ°πÔºåÈÅøÂÖç "Too many pings" ÈîôËØØ
        options = [
            ('grpc.keepalive_time_ms', 300000),  # 5ÂàÜÈíüÔºåÂáèÂ∞ë ping È¢ëÁéá
            ('grpc.keepalive_timeout_ms', 20000),  # 20ÁßíË∂ÖÊó∂
            ('grpc.keepalive_permit_without_calls', False),  # Ê≤°ÊúâË∞ÉÁî®Êó∂‰∏çÂèëÈÄÅ ping
            ('grpc.http2.max_pings_without_data', 2),  # ÂÖÅËÆ∏ÊúÄÂ§ö2‰∏™ ping
            ('grpc.http2.min_time_between_pings_ms', 60000),  # ping ‰πãÈó¥Ëá≥Â∞ëÈó¥Èöî60Áßí
        ]
        
        with grpc.insecure_channel(self.address, options=options) as channel:
            stub = bazi_core_pb2_grpc.BaziCoreServiceStub(channel)
            try:
                response = stub.CalculateBazi(request, timeout=self.timeout)
                
                # ËΩ¨Êç¢‰∏∫Â≠óÂÖ∏Ê†ºÂºè
                result: Dict[str, Any] = {}
                
                # Âü∫Êú¨‰ø°ÊÅØÔºàÂÆâÂÖ®Âú∞ËΩ¨Êç¢‰∏∫Â≠óÂÖ∏ÔºåÈúÄË¶ÅÂèçÂ∫èÂàóÂåñ lunar_dateÔºâ
                basic_info_dict = {}
                if response.basic_info:
                    for key, value in response.basic_info.items():
                        if key == "lunar_date" and isinstance(value, str):
                            # lunar_date ÊòØ JSON Â≠óÁ¨¶‰∏≤ÔºåÈúÄË¶ÅÂèçÂ∫èÂàóÂåñ
                            try:
                                basic_info_dict[key] = json.loads(value) if value else {}
                            except (json.JSONDecodeError, TypeError):
                                basic_info_dict[key] = {}
                        else:
                            basic_info_dict[key] = value
                result["basic_info"] = basic_info_dict
                
                # ÂõõÊü±‰ø°ÊÅØÔºàÂÆâÂÖ®Âú∞Ëé∑ÂèñÔºâ
                result["bazi_pillars"] = {
                    "year": {
                        "stem": response.year_pillar.stem if response.year_pillar else "",
                        "branch": response.year_pillar.branch if response.year_pillar else "",
                    },
                    "month": {
                        "stem": response.month_pillar.stem if response.month_pillar else "",
                        "branch": response.month_pillar.branch if response.month_pillar else "",
                    },
                    "day": {
                        "stem": response.day_pillar.stem if response.day_pillar else "",
                        "branch": response.day_pillar.branch if response.day_pillar else "",
                    },
                    "hour": {
                        "stem": response.hour_pillar.stem if response.hour_pillar else "",
                        "branch": response.hour_pillar.branch if response.hour_pillar else "",
                    },
                }
                
                # ÂõõÊü±ËØ¶ÊÉÖÔºàÂÆâÂÖ®Âú∞Ëé∑ÂèñÔºâ
                result["details"] = {}
                if response.details:
                    for pillar_name, detail in response.details.items():
                        if detail:
                            result["details"][pillar_name] = {
                                "main_star": detail.main_star if hasattr(detail, 'main_star') else "",
                                "hidden_stars": list(detail.hidden_stars) if hasattr(detail, 'hidden_stars') and detail.hidden_stars else [],
                                "hidden_stems": list(detail.hidden_stems) if hasattr(detail, 'hidden_stems') and detail.hidden_stems else [],
                                "star_fortune": detail.star_fortune if hasattr(detail, 'star_fortune') else "",
                                "self_sitting": detail.self_sitting if hasattr(detail, 'self_sitting') else "",
                                "nayin": detail.nayin if hasattr(detail, 'nayin') else "",
                                "kongwang": detail.kongwang if hasattr(detail, 'kongwang') else "",
                                "deities": list(detail.deities) if hasattr(detail, 'deities') and detail.deities else [],
                            }
                        else:
                            result["details"][pillar_name] = {
                                "main_star": "",
                                "hidden_stars": [],
                                "hidden_stems": [],
                                "star_fortune": "",
                                "self_sitting": "",
                                "nayin": "",
                                "kongwang": "",
                                "deities": [],
                            }
                
                # ÂçÅÁ•ûÁªüËÆ°ÔºàÈúÄË¶ÅÂèçÂ∫èÂàóÂåñ JSON Â≠óÁ¨¶‰∏≤Ôºâ
                ten_gods_stats = {}
                if response.ten_gods_stats:
                    for key, value_json in response.ten_gods_stats.items():
                        try:
                            if isinstance(value_json, str):
                                ten_gods_stats[key] = json.loads(value_json) if value_json else {}
                            else:
                                ten_gods_stats[key] = value_json
                        except (json.JSONDecodeError, TypeError):
                            ten_gods_stats[key] = value_json if value_json else {}
                result["ten_gods_stats"] = ten_gods_stats
                
                # ‰∫îË°å‰ø°ÊÅØÔºàÈúÄË¶ÅÂèçÂ∫èÂàóÂåñ JSON Â≠óÁ¨¶‰∏≤ÊàñPythonÂ≠óÂÖ∏Â≠óÁ¨¶‰∏≤Ôºâ
                elements = {}
                if response.elements:
                    for key, value_json in response.elements.items():
                        # DEBUG: ÊâìÂç∞ÂéüÂßãÂÄº
                        logger.info(f"DEBUG: elements['{key}'] raw value type={type(value_json).__name__}, value={repr(value_json)[:200]}")
                        try:
                            if isinstance(value_json, str):
                                # ÂÖàÂ∞ùËØï ast.literal_evalÔºàÊîØÊåÅPythonÂ≠óÂÖ∏Ê†ºÂºèÔºåÂçïÂºïÂè∑Ôºâ
                                try:
                                    import ast
                                    parsed = ast.literal_eval(value_json) if value_json else {}
                                    logger.info(f"DEBUG: ast.literal_eval success for '{key}': {parsed}")
                                except (ValueError, SyntaxError) as e1:
                                    logger.info(f"DEBUG: ast.literal_eval failed for '{key}': {e1}, trying json.loads...")
                                    # Â¶ÇÊûúÂ§±Ë¥•ÔºåÂ∞ùËØï json.loadsÔºàÊ†áÂáÜJSONÔºåÂèåÂºïÂè∑Ôºâ
                                    parsed = json.loads(value_json) if value_json else {}
                                    logger.info(f"DEBUG: json.loads success for '{key}': {parsed}")
                                # Á°Æ‰øùËß£ÊûêÁªìÊûúÊòØÂ≠óÂÖ∏Ôºå‰∏çÊòØÂ≠óÁ¨¶‰∏≤
                                if isinstance(parsed, dict):
                                    elements[key] = parsed
                                else:
                                    logger.warning(f"DEBUG: parsed result for '{key}' is not dict (type={type(parsed)}), using empty dict")
                                    elements[key] = {}
                            elif isinstance(value_json, dict):
                                # Â¶ÇÊûúÂ∑≤ÁªèÊòØÂ≠óÂÖ∏ÔºåÁõ¥Êé•‰ΩøÁî®
                                logger.info(f"DEBUG: elements['{key}'] is already dict: {value_json}")
                                elements[key] = value_json
                            else:
                                # ÂÖ∂‰ªñÁ±ªÂûãÔºà‰∏çÂ∫îËØ•Âá∫Áé∞ÔºâÔºå‰ΩøÁî®Á©∫Â≠óÂÖ∏
                                logger.warning(f"DEBUG: elements['{key}'] is unexpected type: {type(value_json)}")
                                elements[key] = {}
                        except (json.JSONDecodeError, TypeError, ValueError, SyntaxError) as e:
                            # Ëß£ÊûêÂ§±Ë¥•Ôºå‰ΩøÁî®Á©∫Â≠óÂÖ∏ËÄå‰∏çÊòØ‰øùÁïôÈîôËØØÁöÑÂ≠óÁ¨¶‰∏≤ÂÄº
                            logger.warning(f"Failed to parse elements['{key}']: {repr(value_json)[:100]}, error: {e}")
                            elements[key] = {}
                result["elements"] = elements
                
                # ‰∫îË°åËÆ°Êï∞ÔºàÁõ¥Êé•ÊòØ int32Ôºâ
                result["element_counts"] = dict(response.element_counts) if response.element_counts else {}
                
                # ÂÖ≥Á≥ª‰ø°ÊÅØÔºàÈúÄË¶ÅÂèçÂ∫èÂàóÂåñ JSON Â≠óÁ¨¶‰∏≤Ôºâ
                relationships = {}
                if response.relationships:
                    for key, value_json in response.relationships.items():
                        try:
                            if isinstance(value_json, str):
                                relationships[key] = json.loads(value_json) if value_json else {}
                            else:
                                relationships[key] = value_json
                        except (json.JSONDecodeError, TypeError):
                            relationships[key] = value_json if value_json else {}
                result["relationships"] = relationships
                
                # ÂÖÉÊï∞ÊçÆ
                if response.metadata_json:
                    result["metadata"] = json.loads(response.metadata_json)
                
                return result
                
            except grpc.RpcError as e:
                import datetime
                error_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
                logger.error(f"[{error_time}] ‚ùå bazi-core-service (gRPC): Ë∞ÉÁî®Â§±Ë¥• - {e}")
                raise

    def health_check(self) -> bool:
        """ÂÅ•Â∫∑Ê£ÄÊü•"""
        request = bazi_core_pb2.HealthCheckRequest()
        try:
            with grpc.insecure_channel(self.address) as channel:
                stub = bazi_core_pb2_grpc.BaziCoreServiceStub(channel)
                response = stub.HealthCheck(request, timeout=5.0)
                return response.status == "ok"
        except grpc.RpcError:
            logger.exception("bazi-core-service health check failed")
            return False

